File: protenium-iro-hub/app/src/main/java/com/protenium/irohub/ui/start_up/auth/fragment/LoginFragment.java
package com.protenium.irohub.ui.start_up.auth.fragment;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;

import com.protenium.irohub.R;
import com.protenium.irohub.base.BaseFragment;
import com.protenium.irohub.shared_pref.SharedPrefs;
import com.protenium.irohub.ui.main.dashboard.activity.HomeActivity;
import com.protenium.irohub.ui.start_up.auth.viewmodel.AuthViewModel;

import butterknife.BindView;
import butterknife.ButterKnife;


public class LoginFragment extends BaseFragment implements View.OnClickListener {

    private AuthViewModel authViewModel;

    @BindView(R.id.signup)
    TextView signup;

    @BindView(R.id.tvskip)
    TextView tvskip;

    @BindView(R.id.tvForgotPassword)
    TextView tvForgotPassword;

    @BindView(R.id.signin)
    Button signin;

    @BindView(R.id.username)
    EditText etEmailOrNumber;

    @BindView(R.id.password)
    EditText etPassword;

    public static LoginFragment newInstance() {
        Bundle args = new Bundle();
        LoginFragment fragment = new LoginFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View setView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_login, container, false);
        ButterKnife.bind(this, view);
        listener.clearBackStack();
        authViewModel = ViewModelProviders.of((AppCompatActivity) context).get(AuthViewModel.class);

        authViewModel.getIsLoading().observe(this, isLoading -> {
            if (isLoading) {
                showProgress();
            } else {
                dismissProgress();
            }
        });
        signup.setOnClickListener(this);
        tvskip.setOnClickListener(this);
        tvForgotPassword.setOnClickListener(this);
        signin.setOnClickListener(this);

        return view;
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public void onClick(View v) {
        switch (v.getId()) {

            case R.id.signup: {
                listener.replaceFragment(SignUpFragment.newInstance());
                break;
            }
            case R.id.tvskip: {
                break;
            }

            case R.id.tvForgotPassword: {
                listener.replaceFragment(ForgotPasswordFragment.newInstance());
                break;
            }

            case R.id.signin: {
                checkFields();
                break;

            }


        }
    }

    private void checkFields() {
        if (TextUtils.isEmpty(etEmailOrNumber.getText())) {
            etEmailOrNumber.setError("Email is required");
        } else if (TextUtils.isEmpty(etPassword.getText())) {
            etPassword.setError("Password is required");
        } else {
            customerLogin();
        }
    }

    private void customerLogin() {
        authViewModel.userLogin(etEmailOrNumber.getText().toString(), etPassword.getText().toString(), "we").observe(this, loginResponse -> {
            if (loginResponse.getStatus()) {
                SharedPrefs.setBoolean(SharedPrefs.Keys.IS_LOGIN, true);
                SharedPrefs.setString(SharedPrefs.Keys.CUSTOMER_TOKEN, loginResponse.getUser().getAccessToken());
                SharedPrefs.setString(SharedPrefs.Keys.CUSTOMER_ID, loginResponse.getUser().getId().toString());
                SharedPrefs.setString(SharedPrefs.Keys.CUSTOMER_EMAIL, loginResponse.getUser().getEmail());
                Intent intent = new Intent(requireContext(), HomeActivity.class);
                startActivity(intent);
                requireActivity().finishAffinity();

                Toast.makeText(getContext(), "" + loginResponse.getMessage(), Toast.LENGTH_SHORT).show();
            } else {
                Toast.makeText(getContext(), "" + loginResponse.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: joyal670-propertyagent-38e114e5825b/app/src/main/java/com/property/propertyagent/start_up/auth/fragment/LoginFragment.kt
package com.property.propertyagent.start_up.auth.fragment

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Process
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.core.content.ContextCompat
import com.google.firebase.messaging.FirebaseMessaging
import com.property.propertyagent.R
import com.property.propertyagent.agent_panel.ui.main.home.home.HomeActivity
import com.property.propertyagent.base.BaseFragment
import com.property.propertyagent.databinding.FragmentLoginBinding
import com.property.propertyagent.dialogs.InternetDialogFragmentCopy
import com.property.propertyagent.owner_panel.ui.main.home.home.activity.HomeOwnerActivity
import com.property.propertyagent.start_up.auth.viewmodel.LoginViewModel
import com.property.propertyagent.utils.AppPreferences.isLogin
import com.property.propertyagent.utils.AppPreferences.login_type
import com.property.propertyagent.utils.AppPreferences.token
import com.property.propertyagent.utils.AppPreferences.user_email
import com.property.propertyagent.utils.AppPreferences.user_id
import com.property.propertyagent.utils.AppPreferences.user_name
import com.property.propertyagent.utils.AppPreferences.user_phone
import com.property.propertyagent.utils.AppPreferences.user_profileImg
import com.property.propertyagent.utils.Status
import com.property.propertyagent.utils.Toaster
import com.property.propertyagent.utils.isConnected
import com.property.propertyagent.utils.replaceFragment

class LoginFragment : BaseFragment() {

    private lateinit var loginViewModel: LoginViewModel
    private lateinit var binding: FragmentLoginBinding

    private var doubleBackToExitPressedOnce = false
    private var userType: String = ""

    override fun setView(
        inflater: LayoutInflater?,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentLoginBinding.inflate(inflater!!, container, false)
        return binding.root
    }

    override fun initData() {
        when (binding.segBtn.position) {
            0 -> userType = "0"
            1 -> userType = "1"
        }

        requireActivity().onBackPressedDispatcher
            .addCallback(this, object : OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    if (doubleBackToExitPressedOnce) {

                        val pid = Process.myPid()
                        Process.killProcess(pid)

                        val intent = Intent(Intent.ACTION_MAIN)
                        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                        requireContext().startActivity(intent)
                        requireActivity().finish()
                        return
                    }
                    doubleBackToExitPressedOnce = true
                    Toast.makeText(
                        requireContext(),
                        "Please click BACK again to exit",
                        Toast.LENGTH_SHORT
                    ).show()
                    Handler(Looper.getMainLooper()).postDelayed({
                        doubleBackToExitPressedOnce = false
                    }, 5000)
                }
            })
    }

    override fun setupUI() {
    }

    override fun setupViewModel() {
        loginViewModel = LoginViewModel()
    }

    override fun setupObserver() {
        loginViewModel.loginData.observe(this) {
            when (it.status) {
                Status.SUCCESS -> {
                    dismissProgress()

                    /* Saving data to preference */
                    isLogin = true
                    token = it.data?.data?.api_token.toString()
                    login_type = it.data?.data?.type.toString()
                    user_id = it.data!!.data.id
                    user_name = it.data.data.name
                    user_email = it.data.data.email
                    user_phone = it.data.data.phone
                    user_profileImg = it.data.data.profile_image

                    /* start activity after successfully login */
                    if (login_type == "1") {
                        startActivity(Intent(requireActivity(), HomeOwnerActivity::class.java))
                        activity?.finishAffinity()
                    } else if (login_type == "0") {
                        startActivity(Intent(requireActivity(), HomeActivity::class.java))
                        activity?.finishAffinity()
                    }
                }
                Status.LOADING -> {
                    showProgress()
                }
                Status.DATA_EMPTY -> {
                    dismissProgress()
                }
                Status.NO_INTERNET -> {
                    dismissProgress()
                    if (requireContext().isConnected) {
                        Toaster.showToast(
                            requireContext(), getString(R.string.something_wrong),
                            Toaster.State.ERROR, Toast.LENGTH_LONG
                        )
                    } else {
                        val dialog = InternetDialogFragmentCopy(requireActivity())
                        dialog.show(parentFragmentManager, "TAG")
                    }
                }
                Status.ERROR -> {
                    dismissProgress()
                    Toaster.showToast(
                        requireContext(),
                        it.message!!,
                        Toaster.State.ERROR,
                        Toast.LENGTH_LONG
                    )
                }
            }
        }
    }

    override fun onClicks() {
        binding.tvForgot.setOnClickListener {
            appCompatActivity?.replaceFragment(
                fragment = ForgotPasswordFragment.newInstance(binding.segBtn.position, false),
                addToBackStack = true
            )
        }

        binding.tvloginbtn.setOnClickListener {
            val email = binding.fragmentLoginEmailEtx.text.toString()
            val password = binding.fragmentLoginPasswordEtx.text.toString()
            if (email.isBlank() || password.isBlank()) {
                if (email.isBlank()) {
                    binding.tilEmail.error = getString(R.string.email_is_required)
                }
                if (password.isBlank()) {
                    binding.tilPassword.error =
                        getString(R.string.password_is_required)
                }
            } else {
                FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
                    if(task.isSuccessful){
                        loginViewModel.login(
                            email, password, userType, task.result!!
                        )
                    }
                }
            }
        }

        binding.segBtn.setOnPositionChangedListener { position ->
            if (position == 1) {
                binding.segBtn.setSelectedBackground(Color.parseColor("#2F80ED"))
                binding.tvloginbtn.backgroundTintList = context?.let {
                    ContextCompat.getColorStateList(
                        it,
                        R.color.color_accent_blue_statusbar
                    )
                }
                binding.frameLogin.background = context?.let {
                    ContextCompat.getDrawable(
                        it,
                        R.drawable.ic_loginbgblue
                    )
                }
                userType = "1"
            } else {
                binding.segBtn.setSelectedBackground(Color.parseColor("#6AC58C"))
                binding.tvloginbtn.backgroundTintList =
                    context?.let {
                        ContextCompat.getColorStateList(
                            it,
                            R.color.color_accent_green
                        )
                    }
                binding.frameLogin.background = context?.let {
                    ContextCompat.getDrawable(
                        it,
                        R.drawable.ic_loginbg
                    )
                }
                userType = "0"
            }
        }
    }
}